File: ga/glm4_result_2024_6/result/tree_l_glm/49_tree_l.c
use std::io::{self, BufRead};

#[derive(Debug)]
struct TreeNode {
    element: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

impl TreeNode {
    fn new(element: i32) -> Self {
        TreeNode {
            element,
            left: None,
            right: None,
        }
    }

    fn insert(&mut self, element: i32) {
        if element < self.element {
            self.left.get_or_insert_with(|| Box::new(TreeNode::new(element)));
        } else if element > self.element {
            self.right.get_or_insert_with(|| Box::new(TreeNode::new(element)));
        }
    }

    fn delete(&mut self, element: i32) -> bool {
        if self.element == element {
            if self.left.is_none() && self.right.is_none() {
                return true;
            } else if self.left.is_none() {
                *self = *self.right.take().unwrap();
            } else if self.right.is_none() {
                *self = *self.left.take().unwrap();
            } else {
                let mut min_right = self.right.as_mut().unwrap();
                while min_right.left.is_some() {
                    min_right = min_right.left.as_mut().unwrap();
                }
                self.element = min_right.element;
                self.right.as_mut().unwrap().delete(min_right.element);
            }
            true
        } else if element < self.element {
            self.left.as_mut().map(|left| left.delete(element)).unwrap_or(false)
        } else {
            self.right.as_mut().map(|right| right.delete(element)).unwrap_or(false)
        }
    }

    fn find_min(&self) -> Option<&Self> {
        self.left.as_ref().and_then(|left| left.find_min()).or(Some(self))
    }

    fn find_max(&self) -> Option<&Self> {
        self.right.as_ref().and_then(|right| right.find_max()).or(Some(self))
    }

    fn delete_even_numbers(&mut self) {
        if self.element % 2 == 0 {
            self.delete(self.element);
        }
        self.left.as_mut().map(|left| left.delete_even_numbers());
        self.right.as_mut().map(|right| right.delete_even_numbers());
    }

    fn print_in_order(&self) {
        self.left.as_ref().map(|left| left.print_in_order());
        println!("{}", self.element);
        self.right.as_ref().map(|right| right.print_in_order());
    }
}

fn main() {
    let stdin = io::stdin();
    let mut tree: Option<Box<TreeNode>> = None;

    for line in stdin.lock().lines() {
        let line = line.unwrap();
        for token in line.split_whitespace() {
            if let Ok(element) = token.parse::<i32>() {
                if tree.is_none() {
                    tree = Some(Box::new(TreeNode::new(element)));
                } else {
                    tree.as_mut().unwrap().insert(element);
                }
            }
        }
    }

    if let Some(ref mut tree) = tree {
        println!("Tree after insertion:");
        tree.print_in_order();

        tree.delete_even_numbers();

        println!("Tree after deleting even numbers:");
        tree.print_in_order();

        if let Some(min) = tree.find_min() {
            println!("Min is {}, Max is {}", min.element, tree.find_max().unwrap().element);
        }
    } else {
        println!("Tree is empty");
    }
}

File: ga/4omini_result_2024_7/result/tree_l_mini/49_tree_l.rs
use std::cmp::Ord;
use std::io::{self, BufRead};
use std::ptr;
use std::vec::Vec;

#[derive(Debug)]
struct TreeNode {
    element: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

struct SearchTree {
    root: Option<Box<TreeNode>>,
}

impl SearchTree {
    fn new() -> Self {
        SearchTree { root: None }
    }

    fn insert(&mut self, x: i32) {
        self.root = Self::insert_node(self.root.take(), x);
    }

    fn insert_node(node: Option<Box<TreeNode>>, x: i32) -> Option<Box<TreeNode>> {
        match node {
            None => Some(Box::new(TreeNode {
                element: x,
                left: None,
                right: None,
            })),
            Some(mut n) => {
                if x < n.element {
                    n.left = Self::insert_node(n.left, x);
                } else if x > n.element {
                    n.right = Self::insert_node(n.right, x);
                }
                Some(n)
            }
        }
    }

    fn delete_even_numbers(&mut self) {
        self.root = Self::delete_even_node(self.root.take());
    }

    fn delete_even_node(node: Option<Box<TreeNode>>) -> Option<Box<TreeNode>> {
        match node {
            None => None,
            Some(mut n) => {
                n.left = Self::delete_even_node(n.left);
                n.right = Self::delete_even_node(n.right);

                if n.element % 2 == 0 {
                    return Self::join(n.left, n.right);
                }
                Some(n)
            }
        }
    }

    fn join(left: Option<Box<TreeNode>>, right: Option<Box<TreeNode>>) -> Option<Box<TreeNode>> {
        match (left, right) {
            (None, None) => None,
            (Some(l), None) => Some(l),
            (None, Some(r)) => Some(r),
            (Some(l), Some(r)) => {
                let min_node = Self::find_min(&r);
                let new_root = min_node.clone();
                let new_right = Self::delete_node(r, new_root.element);
                Some(Box::new(TreeNode {
                    element: new_root.element,
                    left: Some(l),
                    right: new_right,
                }))
            }
        }
    }

    fn find_min(node: &Option<Box<TreeNode>>) -> &Box<TreeNode> {
        match node {
            Some(n) if n.left.is_some() => Self::find_min(&n.left),
            Some(n) => n,
            None => panic!("Tree is empty"),
        }
    }

    fn delete_node(node: Option<Box<TreeNode>>, x: i32) -> Option<Box<TreeNode>> {
        match node {
            None => return None,
            Some(mut n) => {
                if x < n.element {
                    n.left = Self::delete_node(n.left, x);
                } else if x > n.element {
                    n.right = Self::delete_node(n.right, x);
                } else {
                    // Node with the value x found
                    if n.left.is_none() {
                        return n.right;
                    } else if n.right.is_none() {
                        return n.left;
                    } else {
                        let min_node = Self::find_min(&n.right);
                        n.element = min_node.element;
                        n.right = Self::delete_node(n.right, min_node.element);
                    }
                }
                Some(n)
            }
        }
    }

    fn print_in_order(&self) {
        Self::in_order(&self.root);
    }

    fn in_order(node: &Option<Box<TreeNode>>) {
        if let Some(n) = node {
            Self::in_order(&n.left);
            println!("{}", n.element);
            Self::in_order(&n.right);
        }
    }

    fn find_min(&self) -> Option<i32> {
        self.root.as_ref().map(|node| Self::find_min(&Some(node.clone())).element)
    }

    fn find_max(&self) -> Option<i32> {
        self.root.as_ref().map(|node| Self::find_max(&Some(node.clone())).element)
    }

    fn find_max(node: &Option<Box<TreeNode>>) -> &Box<TreeNode> {
        match node {
            Some(n) if n.right.is_some() => Self::find_max(&n.right),
            Some(n) => n,
            None => panic!("Tree is empty"),
        }
    }
}

fn main() {
    let stdin = io::stdin();
    let input = stdin.lock().lines().next().unwrap().unwrap();
    let nums: Vec<i32> = input
        .split_whitespace()
        .filter_map(|token| token.parse::<i32>().ok())
        .collect();

    let mut tree = SearchTree::new();

    for &num in &nums {
        tree.insert(num);
    }

    if nums.is_empty() {
        println!("Tree is empty");
        return;
    }

    println!("Tree after insertion:");
    tree.print_in_order();

    // Delete all even numbers
    tree.delete_even_numbers();

    println!("Tree after deleting even numbers:");
    tree.print_in_order();

    match (tree.find_min(), tree.find_max()) {
        (Some(min), Some(max)) => {
            println!("Min is {}, Max is {}", min, max);
        }
        _ => {
            println!("Tree is empty");
        }
    }
}